src/sleep/sleep.c

```diff
@@ -240,6 +240,52 @@ static int lock_all_homes(void) {
         return 0;
 }
 
+/*
+ * The 'wakeup_count' attribute provides a means by which wakeup events can be
+ * handled in a non-racy way.
+ *
+ * If a wakeup event occurs when the system is in a sleep state, it simply is
+ * woken up.  In turn, if an event that would wake the system up from a sleep
+ * state occurs when it is undergoing a transition to that sleep state, the
+ * transition should be aborted.  Moreover, if such an event occurs when the
+ * system is in the working state, an attempt to start a transition to the
+ * given sleep state should fail during certain period after the detection of
+ * the event.  Using the 'state' attribute alone is not sufficient to satisfy
+ * these requirements, because a wakeup event may occur exactly when 'state'
+ * is being written to and may be delivered to user space right before it is
+ * frozen, so the event will remain only partially processed until the system is
+ * woken up by another event.  In particular, it won't cause the transition to
+ * a sleep state to be aborted.
+ *
+ * For more details see https://lwn.net/Articles/392897/
+ */
+static int write_wakeup_count(void) {
+        static const char kernel_wakeup_count[] = "/sys/power/wakeup_count";
+        static const char systemd_wakeup_count[] = "/run/wakeup_count";
```

Comment:
i don't really see the benefit of these constants. just spell out the strings directly where needed, it's easier to follow